File: vciplreportsv1001/vcipl1001/report/customer_&_supplier_fill_ratio/customer_&_supplier_fill_ratio.js
frappe.query_reports["Customer & Supplier Fill Ratio"] = {

    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company")
        },
        { fieldname: "from_date", label: __("From Date"), fieldtype: "Date" },
        { fieldname: "to_date", label: __("To Date"), fieldtype: "Date" }
    ],

    formatter(value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (column.fieldname === "fill_ratio") {

            if (data.fill_ratio < 50)
                value = `<span style="color:red;font-weight:bold">${data.fill_ratio}%</span>`;
            else if (data.fill_ratio < 80)
                value = `<span style="color:orange;font-weight:bold">${data.fill_ratio}%</span>`;
            else
                value = `<span style="color:green;font-weight:bold">${data.fill_ratio}%</span>`;
        }

        if (column.fieldname === "risk") {

            if (data.risk === "Critical")
                value = `<span style="color:red;font-weight:bold">🔴 Critical</span>`;
            else if (data.risk === "Warning")
                value = `<span style="color:orange;font-weight:bold">🟠 Warning</span>`;
            else
                value = `<span style="color:green;font-weight:bold">🟢 OK</span>`;
        }

        if (column.fieldname === "order_count" && data.order_popup) {

            let safe = encodeURIComponent(data.order_popup);

            value = `<a style="font-weight:bold;color:#1674E0;cursor:pointer"
                onclick='show_orders("${safe}", "${data.party}")'>
                ${data.order_count}
            </a>`;
        }

        return value;
    }
};


// ================= PRINT =================

window.print_section = function(id) {

    let content = document.getElementById(id).innerHTML;

    let w = window.open("", "", "width=1000,height=800");

    w.document.write(`
        <html>
        <head>
            <title>Print</title>
            <style>
                body { font-family: Arial; padding: 20px; }
                table { border-collapse: collapse; width: 100%; }
                th, td { border: 1px solid black; padding: 6px; text-align:center; }
            </style>
        </head>
        <body>${content}</body>
        </html>
    `);

    w.document.close();
    w.print();
};


// ================= LEVEL 2 =================

window.show_orders = function(encoded, party) {

    let rows = JSON.parse(decodeURIComponent(encoded));

    let tot_qty = 0, tot_del = 0;

    let html = `<div id="print_orders">
    <h3>Pending Orders — ${party}</h3>
    <table class="table table-bordered">
    <thead>
        <tr>
            <th>SO No</th>
            <th>Date</th>
            <th>Pending Delivery</th>
            <th>Ordered</th>
            <th>Delivered</th>
            <th>Fill %</th>
        </tr>
    </thead><tbody>`;

    rows.forEach(r => {

        tot_qty += r.qty;
        tot_del += r.delivered;

        let safe = encodeURIComponent(r.pending_popup);

        html += `<tr>
            <td><b>${r.so_no}</b></td>
            <td>${r.so_date}</td>
            <td><a style="color:red;font-weight:bold;cursor:pointer"
                onclick='show_items("${safe}")'>View Pending</a></td>
            <td>${r.qty}</td>
            <td>${r.delivered}</td>
            <td>${r.fill_ratio}%</td>
        </tr>`;
    });

    let total_ratio = tot_qty ? ((tot_del / tot_qty) * 100).toFixed(2) : 0;

    html += `<tr style="font-weight:bold;background:#eee">
        <td colspan="3">Grand Total</td>
        <td>${tot_qty}</td>
        <td>${tot_del}</td>
        <td>${total_ratio}%</td>
    </tr></tbody></table></div>`;

    let d = new frappe.ui.Dialog({
        title: "Order Details",
        size: "extra-large",
        fields: [{ fieldtype: "HTML", fieldname: "html" }]
    });

    d.fields_dict.html.$wrapper.html(`
        ${html}
        <button class="btn btn-primary"
            onclick="print_section('print_orders')">Print</button>
    `);

    d.show();
};


// ================= LEVEL 3 =================

window.show_items = function(encoded) {

    let rows = JSON.parse(decodeURIComponent(encoded));

    let tot_qty = 0, tot_del = 0, tot_pen = 0;

    let html = `<div id="print_items">
    <h3>Pending Items</h3>
    <table class="table table-bordered">
    <thead>
        <tr>
            <th>Item</th>
            <th>Ordered</th>
            <th>Delivered</th>
            <th>Pending</th>
            <th>Fill %</th>
        </tr>
    </thead><tbody>`;

    rows.forEach(r => {

        let ratio = r.qty ? ((r.delivered / r.qty) * 100).toFixed(2) : 0;

        tot_qty += r.qty;
        tot_del += r.delivered;
        tot_pen += r.pending;

        html += `<tr>
            <td>${r.item}</td>
            <td>${r.qty}</td>
            <td>${r.delivered}</td>
            <td style="color:red;font-weight:bold">${r.pending}</td>
            <td>${ratio}%</td>
        </tr>`;
    });

    let total_ratio = tot_qty ? ((tot_del / tot_qty) * 100).toFixed(2) : 0;

    html += `<tr style="font-weight:bold;background:#eee">
        <td>Grand Total</td>
        <td>${tot_qty}</td>
        <td>${tot_del}</td>
        <td>${tot_pen}</td>
        <td>${total_ratio}%</td>
    </tr></tbody></table></div>`;

    let d = new frappe.ui.Dialog({
        title: "Pending Items",
        size: "large",
        fields: [{ fieldtype: "HTML", fieldname: "html" }]
    });

    d.fields_dict.html.$wrapper.html(`
        ${html}
        <button class="btn btn-primary"
            onclick="print_section('print_items')">Print</button>
    `);

    d.show();
};

File: vciplreportsv1001/vcipl1001/report/customer_&_supplier_fill_ratio/customer_&_supplier_fill_ratio.py
import frappe
from frappe import _
from frappe.utils import flt, date_diff


def execute(filters=None):

    validate_filters(filters)

    columns = get_columns()
    data = get_data(filters)

    return columns, data


# ---------------- VALIDATION ----------------

def validate_filters(filters):
    if filters.get("from_date") and filters.get("to_date"):
        if date_diff(filters.to_date, filters.from_date) < 0:
            frappe.throw(_("To Date cannot be before From Date."))


# ---------------- DATA ----------------

def get_data(filters):

    conditions = ""

    if filters.get("company"):
        conditions += " AND company = %(company)s"

    if filters.get("from_date") and filters.get("to_date"):
        conditions += " AND transaction_date BETWEEN %(from_date)s AND %(to_date)s"

    party_type = filters.get("party_type", "Both")
    queries = []

    # ---------- CUSTOMER ----------
    if party_type in ["Both", "Customer"]:
        queries.append(f"""
            SELECT
                'Customer' AS party_type,
                c.customer_name AS party,
                COUNT(DISTINCT so.name) AS order_count,
                SUM(soi.qty) AS qty,
                SUM(soi.delivered_qty) AS delivered_qty,
                SUM(soi.qty - soi.delivered_qty) AS pending_qty
            FROM `tabSales Order` so
            INNER JOIN `tabSales Order Item` soi ON soi.parent = so.name
            INNER JOIN `tabCustomer` c ON c.name = so.customer
            WHERE so.docstatus = 1
            AND so.status NOT IN ('Completed','Closed','Stopped','On Hold')
            {conditions}
            GROUP BY c.customer_name
        """)

    # ---------- SUPPLIER ----------
    if party_type in ["Both", "Supplier"]:
        queries.append(f"""
            SELECT
                'Supplier' AS party_type,
                s.supplier_name AS party,
                COUNT(DISTINCT po.name) AS order_count,
                SUM(poi.qty) AS qty,
                SUM(poi.received_qty) AS delivered_qty,
                SUM(poi.qty - poi.received_qty) AS pending_qty
            FROM `tabPurchase Order` po
            INNER JOIN `tabPurchase Order Item` poi ON poi.parent = po.name
            INNER JOIN `tabSupplier` s ON s.name = po.supplier
            WHERE po.docstatus = 1
            AND po.status NOT IN ('Completed','Closed','Stopped')
            {conditions}
            GROUP BY s.supplier_name
        """)

    raw = frappe.db.sql(" UNION ALL ".join(queries), filters, as_dict=True)

    final = []

    for row in raw:

        # fill ratio
        row.fill_ratio = round(
            (flt(row.delivered_qty) / flt(row.qty) * 100)
            if row.qty else 0,
            2
        )

        # risk classification
        if row.fill_ratio < 50:
            row.risk = "Critical"
        elif row.fill_ratio < 80:
            row.risk = "Warning"
        else:
            row.risk = "OK"

        # risk filter
        rf = filters.get("risk_filter")

        if rf == "HIGH" and row.risk != "Critical":
            continue
        if rf == "Medium" and row.risk != "Warning":
            continue
        if rf == "OK" and row.risk != "OK":
            continue

        final.append(row)

    return final


# ---------------- COLUMNS ----------------

def get_columns():

    return [
        {"label": _("Party Type"), "fieldname": "party_type", "width": 120},
        {"label": _("Customer / Supplier"), "fieldname": "party", "width": 260},

        {"label": _("Order Count"), "fieldname": "order_count", "fieldtype": "Int", "width": 120},

        {"label": _("Total Ordered"), "fieldname": "qty", "fieldtype": "Float"},
        {"label": _("Delivered"), "fieldname": "delivered_qty", "fieldtype": "Float"},
        {"label": _("Pending"), "fieldname": "pending_qty", "fieldtype": "Float"},

        {"label": _("Fill %"), "fieldname": "fill_ratio", "fieldtype": "Percent"},
        {"label": _("Risk Level"), "fieldname": "risk", "width": 120},
    ]
